core/trino-main/src/main/java/io/trino/operator/StreamingDirectExchangeBuffer.java

```diff
@@ -119,7 +122,7 @@ public void addPages(TaskId taskId, List<Slice> pages)
             bufferedPages.addAll(pages);
             bufferRetainedSizeInBytes += pagesRetainedSizeInBytes;
             maxBufferRetainedSizeInBytes = max(maxBufferRetainedSizeInBytes, bufferRetainedSizeInBytes);
-            unblockIfNecessary(blocked);
+            unblockIfNecessary(pages.size());
```

Comment:
Please add a comment about why we're unblocking same number of consumers as pages.